### PR TITLE
Store avatars in database

### LINKED ISSUE #182: Profile Picture does not survive

### PR DESCRIPTION
**Pull request by felixauringer**
Fixes #182 

I removed ActiveStorage because we do not have access to something like S3. This PR saves our avatars as binary objects in the database. That may not be the prettiest solution but it should work on heroku.  
I used [this article](https://gist.github.com/macek/610596) as a guide.

Todos for my next working day:

- [x] update tests
- [x] set default avatar on creation of user
- [x] reload the shown image after update using javascript (reloading the whole page would empty the other form on the same page)
- [x] maybe try to fix the migration history properly (so that rollbacks would be possible)
- [x] remove unnecessary gem
- [x] properly display errors

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -4,14 +4,6 @@ class ApplicationController < ActionController::Base
...
 
   before_action :set_variant
-  before_action :configure_permitted_parameters,
-                if: :devise_controller?
-
-  def configure_permitted_parameters
-    devise_parameter_sanitizer.permit(:account_update) do |u|
-      u.permit(:avatar, :email, :password, :password_confirmation, :current_password)
-    end
-  end
```
Comment (frcroth) [inline app/controllers/application_controller.rb:14]: What's the reason for this change?

File: app/controllers/avatars_controller.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  def upload
+    @avatar = Avatar.find_by(user_id: current_user.id) || Avatar.new
+    file = params[:avatar][:file]
+    @saved = @avatar.update(
+      file: file.read,
+      filename: file.original_filename,
+      filesize: file.size,
+      mime_type: file.content_type,
+      user_id: current_user.id
```
Comment (frcroth) [inline app/controllers/avatars_controller.rb]: Should this be via current_user or rather via params user_id?

File: spec/models/avatar_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  it 'is not valid without a filename' do
+    expect(FactoryBot.build(:avatar, filename: nil)).not_to be_valid
+  end
+
+  it 'is not valid without a filesize' do
+    expect(FactoryBot.build(:avatar, filesize: nil)).not_to be_valid
+  end
+
+  it 'is not valid if it is bigger than 5MiB' do
```
Comment (frcroth) [inline spec/models/avatar_spec.rb:22]: Is it megabytes or mibibyte?

File: app/views/avatars/upload.js.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    <% end %>
+    const target = document.getElementById('new_avatar');
+    if (target.firstChild.id === 'error_explanation') {
+        target.replaceChild(error_node, target.firstChild);
+    } else {
+        target.insertBefore(error_node, target.firstChild);
+    }
+    generatePopup("<%= I18n.t('denial.resource_update', resource: Avatar) %>", 'fa-exclamation', [], 3000);
+  }
+<% end %>
```
Comment (frcroth) [inline app/views/avatars/upload.js.erb:28]: Don't let @sirkrypt0 see this, or he will turn it all into jQuery

File: app/views/avatars/upload.js.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  {
+    const error_node = document.createElement('div');
+    error_node.id = 'error_explanation';
+    const list = document.createElement('ul');
+    error_node.appendChild(list);
+    let list_item;
+    <% @avatar.errors.full_messages.each do |message| %>
+      list_item = document.createElement('li');
+      list_item.textContent = '<%= message %>';
+      list.appendChild(list_item)
```
Comment (frcroth) [inline app/views/avatars/upload.js.erb]:
```suggestion
      list.appendChild(list_item);
```

File: spec/models/avatar_spec.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    expect(FactoryBot.build(:avatar, filesize: 5.megabytes + 1)).not_to be_valid
+  end
+
+  it 'is not valid with a rational filesize' do
```
Comment (frcroth) [inline spec/models/avatar_spec.rb]: What's a rational file size?

File: app/models/avatar.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# A model for storing user avatars in the database
+class Avatar < ApplicationRecord
+  VALID_MIME_TYPES = %w[image/png image/jpg image/jpeg image/gif].freeze
+  MAX_FILE_SIZE_MEGABYTES = 5
+
+  validates :filename, :filesize, :file, :mime_type, presence: true
```
Comment (frcroth) [inline app/models/avatar.rb]:
```suggestion
  validates :filename, :file_size, :file, :mime_type, presence: true
```
And everywhere else

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (felixauringer) [reply app/controllers/application_controller.rb:14]: This was added in the [PR](https://github.com/hpi-swt2/connections-portal/pull/156/files) that introduced profile pictures. I think that everything but the avatar is permitted by default because devise needs it. And because we do not need the avatar anymore I completely removed it to fall back to the default behaviour again.

File: app/controllers/avatars_controller.rb (hunk 2 above)
Comment (felixauringer) [reply app/controllers/avatars_controller.rb]: There shouldn't be a `params[:user_id]` because avatars are not a subresource of the user. But because one should only change the profile picture of oneself, `current_user` is also the desired behaviour here.

File: spec/models/avatar_spec.rb (hunk 3 above)
Comment (felixauringer) [reply spec/models/avatar_spec.rb:22]: it does not return `5,000,000` but a little bit more, so I think it is actually M**e**biByte.

Seems like the [rails naming](https://api.rubyonrails.org/classes/Numeric.html#method-i-megabyte) isn't right here

File: app/views/avatars/upload.js.erb (hunk 4 above)
Comment (felixauringer) [reply app/views/avatars/upload.js.erb:28]: I know this is ugly but I didn't know how to do it better. If jQuery is the way to go, then he can of course refactor this ^^

File: spec/models/avatar_spec.rb (hunk 6 above)
Comment (felixauringer) [reply spec/models/avatar_spec.rb]: A non-integer one, what would be the appropriate wording to describe this?

File: app/models/avatar.rb (hunk 7 above)
Comment (felixauringer) [reply app/models/avatar.rb]: Should I do the same for the `filename`?

Comment (frcroth) [reply app/models/avatar.rb]: No, filename is an English word https://www.oxfordlearnersdictionaries.com/definition/english/filename?q=filename

**Comment by frcroth** [comment]
Is there a way to access the avatars easily via a link that is predictable without ruby (for the js part of the global chat)? For #179

File: spec/models/avatar_spec.rb (hunk 3 above)
Comment (frcroth) [reply spec/models/avatar_spec.rb:22]: 5,243e+6 Bytes is 5 Mebibytes, if this is the returned value, change all occurrences of megabytes to mebibytes.

File: spec/models/avatar_spec.rb (hunk 6 above)
Comment (frcroth) [reply spec/models/avatar_spec.rb]: Ah no I get it, you mean as in rational number

Comment (frcroth) [inline spec/models/avatar_spec.rb]:
```suggestion
  it 'is not valid with a non integer filesize' do
```

File: spec/models/avatar_spec.rb (hunk 3 above)
Comment (felixauringer) [reply spec/models/avatar_spec.rb:22]: `.megabytes` is a rails method, I could write an alias to that and name it `mebibytes` but I'm not sure if that would be a good solution. I could not find other (non-method calls) occurences of it in my code, did you?

**Comment by felixauringer** [comment]
> Is there a way to access the avatars easily via a link that is predictable without ruby (for the js part of the global chat)? For #179

As the id of the avatar is unrelated to anything else, that is not possible. The easieast way I can think of is providing the link to the avatar together with the rest of the information (message, user, ...) via websocket. In my opinion, building links in javascript shouldn't be done anyway because of the many places one would have to change something if the link format would change.

File: app/models/avatar.rb (hunk 7 above)
Comment (felixauringer) [reply app/models/avatar.rb]: Done

File: spec/models/avatar_spec.rb (hunk 3 above)
Comment (felixauringer) [reply spec/models/avatar_spec.rb:22]: I did find one (the constant in the avatar model) and changed it there

File: spec/models/avatar_spec.rb (hunk 6 above)
Comment (felixauringer) [reply spec/models/avatar_spec.rb]: That's much clearer with your suggestion now

**Review by frcroth — APPROVED** [review]

**Review by konradh — APPROVED** [review]

**Comment by konradh** [comment]
> > Is there a way to access the avatars easily via a link that is predictable without ruby (for the js part of the global chat)? For #179
> 
> As the id of the avatar is unrelated to anything else, that is not possible. The easieast way I can think of is providing the link to the avatar together with the rest of the information (message, user, ...) via websocket. In my opinion, building links in javascript shouldn't be done anyway because of the many places one would have to change something if the link format would change.

We moved the avatars to `/users/#{id}/avatar`.

File: app/views/users/edit_profile.html.erb [hunk 8]
```diff
@@ -1,17 +1,23 @@
 <h2><%= I18n.t('user.edit.edit_profile_settings_heading') %></h2>
 
-<%= form_with(model: @user, local: true) do |form| %>
-  <% if @user.errors.any? %>
-    <%= render 'layouts/flash_messages'%>
-  <% end %>
+<p>
+  <%= render 'avatar', user: @user, size: 160, classes: %i[rounded-circle centered-avatar] %>
```
Comment (Paula-Kli) [inline app/views/users/edit_profile.html.erb:4]: should `centered-avatar` be added to the `card.html` as well?

**Review by Paula-Kli — APPROVED** [review]

File: app/views/users/edit_profile.html.erb (hunk 8 above)
Comment (felixauringer) [reply app/views/users/edit_profile.html.erb:4]: It wasn't there before so I assumed that it is not necessary there. In my opinion it could be omitted completely as it does not look good on the profile page but this PR was more about making it work again, not making it pretty :sweat_smile:
